github-actions/create-pr-for-changes/lib/main.ts

```diff
@@ -0,0 +1,204 @@
+      core.info(
+        `Skipping PR creation, because there is already a PR: #${matchingPrs[0].number} ` +
+          `(${matchingPrs[0].html_url})`,
+      );
+      return;
+    } else {
+      core.info(`No pre-existing PR found for branch '${branchName}'.`);
+    }
+
+    // Gather PRs (if any) that will be superseded by the PR that will be created.
+    const {
+      data: {items: supersededPrs},
+    } = await git.github.search.issuesAndPullRequests({
+      q: `is:pull-request is:open repo:${owner}/${repo} base:${base} head:${branchPrefix}-`,
+      // NOTE:
+      //   We assume that there will be no more than 100 open PRs for the same type of change. This
+      //   is a reasonable assumption for our purposes, but even if it doesn't hold it won't be a
+      //   big problem: It just means that older PRs will not be auto-closed and will not be marked
+      //   as superseded.
+      per_page: 100,
+      sort: 'created',
+      order: 'desc',
+    });
+
+    core.info(
+      `Found ${supersededPrs.length} pre-existing PR(s) that will be superseded: ` +
+        (supersededPrs.length === 0 ? '-' : supersededPrs.map((pr) => `#${pr.number}`).join(', ')),
+    );
+
+    // Commit the current changes on a new branch.
+    const prTitle = core.getInput('pr-title', {required: true});
+    const prDescription = core.getInput('pr-description', {required: true});
+    const prFooter =
+      supersededPrs.length === 0
+        ? ''
+        : `\n${supersededPrs.map((pr) => `\nCloses #${pr.number}`).join('')}`;
+    const prBody = `${prDescription}${prFooter}`;
+    const prLabels = core
+      .getInput('pr-labels', {required: false})
+      .split('\n')
+      .map((l) => l.trim())
+      .filter((l) => l !== '');
+    const commitMessage = `${prTitle}\n\n${prBody}`;
+
+    git.run(['checkout', '-b', branchName]);
+    git.run(['add', '--all']);
+    git.run(['commit', '-m', commitMessage, '--no-verify']);
+    git.run(['push', '--force', git.getRepoGitUrl(), `HEAD:refs/heads/${branchName}`]);
+
+    core.info('Committed changes.');
+
+    // Create a PR and add the specified labels (if any).
+    const {data: createdPr} = await git.github.pulls.create({
+      owner,
+      repo,
+      base,
+      head: branchName,
+      title: prTitle,
+      body: prBody,
+      maintainer_can_modify: true,
+    });
+
+    core.setOutput('result', ActionResult.created);
+    core.setOutput('pr-number', `${createdPr.number}`);
+
+    core.info(`Created PR #${createdPr.number} (${createdPr.html_url}).`);
+
+    if (prLabels.length > 0) {
+      await git.github.issues.addLabels({
+        owner,
+        repo,
+        issue_number: createdPr.number,
+        labels: prLabels,
+      });
+
+      core.info(`Added ${prLabels.length} label(s) to PR: '${prLabels.join("', '")}'`);
+    }
+
+    // Comment on superseded PRs.
+    if (supersededPrs.length > 0) {
+      for (const pr of supersededPrs) {
+        await git.github.issues.createComment({
```

Comment:
This should be closing the PR right not just noting it superceded, or at least we should support that as an option.

Reply:
My reasoning for not closing the PRs directly is to account for situations like the following:
- PR #1 is created which has some changes.
- PR #1 passes CI checks, looks good for merging, but is not yet merged.
- On the next run, PR #2 is created which has more changes than PR #1.
- It turns out that the changes in PR #2 are not valid and cause failures.
- At this point, we might want to get PR #1 in (which includes useful updates), while we address the issues that cause PR #2 to fail.

All superseded PRs are referenced in the superseding PR's commit message and will be automatically closed once the superseding PR is merged.

##
I personally like this workflow, but I understand it might be confusing for people that are not familiar with how the action works. Happy to change this to close the superseded PRs instead (or support both options).

WDYT?

Reply:
That makes sense, I think I will need to see it in action before I can really judge it. What you laid out makes sense, I think I worry about how deep of superseding we could end up with.  Lets go with this and if we determine that its better to close we couldn't be able to do that.